File: Python/MergeSort.py
# importing the sys module 
import sys 
  
# the setrecursionlimit function is 
# used to modify the default recursion 
# limit set by python. Using this,  
# we can increase the recursion limit 
# to satisfy our needs 

sys.setrecursionlimit(10**6) 


def merge(arr,p,q,r):
    # q is the middle element 
    n1 = q-p+1
    n2 = r-q
    left = []
    right = []
    i = 0
    j = 0
    # divide the array into left and right sub-parts
    for k in range(n1):
        left.append(arr[p+k])
    left.append(12345678)
    # appending a big-integer as there might be a case where there is no elements in the left
    # or the right sub-array ... so it means at that case all the elements has to be copied to the
    # given arr 
    for k in range(n2):
        right.append(arr[q+k])
    right.append(12345678)
    #compare left and right sub-array
    for k in range(p,r):
        if left[i] <= right[j]:
            arr[k] = left[i]
            i = i + 1
        else:
            arr[k] = right[j]
            j = j + 1

def mergesort(arr,p,r):
    if(p<r):
        q = p + (r-p)//2
        mergesort(arr,p,q)
        mergesort(arr,q+1,r)
        merge(arr,p,q,r)

if __name__ == '__main__':
    print("ENTER UNSORTED ARRAY: ")
    arr = list(map(int,input().split()))
    mergesort(arr,0,(len(arr)-1))
    print("SORTED ARRAY :")
    print(arr)

File: C/mergesort.c
#include <stdio.h>
void merge(int arr[],int low,int mid,int high)
{
	int n1 = mid - low + 1;
	int n2 = high - mid;
	int l[n1+1],r[n2+1];
	for(int i = 1; i<=n1; i++)
	{
		l[i] = arr[low + i-1];
	}
	for(int j=1;j <=n2; j++)
	{
		r[j] = arr[mid + j];
	}
	l[n1 + 1] = 99999;
	r[n2 + 1] = 99999;
	int i = 1;
	int j = 1;
	for(int k = low; k <=high; k++)
	{
		if(l[i]<r[j])
		{
			arr[k] = l[i];
			i++;
		}
		else
		{
			arr[k] = r[j];
			j++;
		}
	}
}

void mergesort(int arr[],int low,int high)
{
	if(low<high)
	{
		int q = low + (high - low)/2;
		mergesort(arr, low, q);
		mergesort(arr, q+1, high);
		merge(arr, low, q, high); 
	}	
}

void print(int arr[],int n)
{
	for(int i=0;i<n;i++)
	{
		printf("%d ",arr[i]);
	}
}

int main()
{
	int arr[] = {1,4,7,6,8,5,3,0,2};
	print(arr,9);
	mergesort(arr, 0, 8);
	printf("\n");
	print(arr,9);
  return 0;
}
